go/vt/topo/helpers/compare.go

```diff
@@ -0,0 +1,261 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package helpers contains a few utility classes to handle topo.Server
+// objects, and transitions from one topo implementation to another.
+package helpers
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"golang.org/x/net/context"
+	"vitess.io/vitess/go/vt/concurrency"
+	"vitess.io/vitess/go/vt/topo"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+)
+
+// CompareKeyspaces will create the keyspaces in the destination topo.
+func CompareKeyspaces(ctx context.Context, fromTS, toTS *topo.Server) error {
+	keyspaces, err := fromTS.GetKeyspaces(ctx)
+	if err != nil {
+		return fmt.Errorf("GetKeyspace(%v): %v", keyspaces, err)
+	}
+
+	wg := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+	for _, keyspace := range keyspaces {
+		wg.Add(1)
+		go func(keyspace string) {
+			defer wg.Done()
+
+			fromKs, err := fromTS.GetKeyspace(ctx, keyspace)
+			if err != nil {
+				rec.RecordError(fmt.Errorf("GetKeyspace(%v): %v", keyspace, err))
+				return
+			}
+
+			toKs, err := toTS.GetKeyspace(ctx, keyspace)
+			if err != nil {
+				rec.RecordError(fmt.Errorf("GetKeyspace(%v): %v", keyspace, err))
+				return
+			}
+
+			if !reflect.DeepEqual(fromKs.Keyspace, toKs.Keyspace) {
+				rec.RecordError(fmt.Errorf("Keyspace: %v does not match between from and to topology", keyspace))
+				return
+			}
+
+			fromVs, err := fromTS.GetVSchema(ctx, keyspace)
+			switch {
+			case err == nil:
+				// Nothing to do.
+			case topo.IsErrType(err, topo.NoNode):
+				// Nothing to do.
+			default:
+				rec.RecordError(fmt.Errorf("GetVSchema(%v): %v", keyspace, err))
+				return
+			}
+
+			toVs, err := toTS.GetVSchema(ctx, keyspace)
+			switch {
+			case err == nil:
+				// Nothing to do.
+			case topo.IsErrType(err, topo.NoNode):
+				// Nothing to do.
+			default:
+				rec.RecordError(fmt.Errorf("GetVSchema(%v): %v", keyspace, err))
+				return
+			}
+
+			if !reflect.DeepEqual(fromVs, toVs) {
+				rec.RecordError(fmt.Errorf("Vschema for keyspace: %v does not match between from and to topology", keyspace))
+				return
+			}
+		}(keyspace)
+	}
+	wg.Wait()
+	if rec.HasErrors() {
+		return fmt.Errorf("compareKeyspaces failed: %v", rec.Error())
+	}
+	return nil
+}
+
+// CompareShards will create the shards in the destination topo.
+func CompareShards(ctx context.Context, fromTS, toTS *topo.Server) error {
+	keyspaces, err := fromTS.GetKeyspaces(ctx)
+	if err != nil {
+		return fmt.Errorf("fromTS.GetKeyspaces: %v", err)
+	}
+
+	wg := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+	for _, keyspace := range keyspaces {
+		wg.Add(1)
+		go func(keyspace string) {
+			defer wg.Done()
+			shards, err := fromTS.GetShardNames(ctx, keyspace)
+			if err != nil {
+				rec.RecordError(fmt.Errorf("GetShardNames(%v): %v", keyspace, err))
+				return
+			}
+
+			for _, shard := range shards {
+				wg.Add(1)
+				go func(keyspace, shard string) {
```

Comment:
I think we would be well-served to _reduce_ the parallelism in the compare, both for this and for the tablet case below.

I'm concerned this will put unnecessarily spiky load on the topo service and we might run into the rate limits or (worse) cause downstream problems just for the purpose of doing the comparison faster.

FWIW I feel similarly about the already-existing copy helpers -- it seems much better to be if we do these kind of rare topology bulk operations sequentially rather than slam a ton of goroutines at the topo service at once.